### PR TITLE
Cluster scaling

### PR DESCRIPTION
**Pull request by russgold**
This resolves issue https://jira.oraclecorp.com/jira/browse/OWLS-72885

I have refactored access to the service fields in DomainPresenceInfo and ServerKubernetesObjects, so that now all accesses should go through ServiceHelper.

### CODE REVIEW

**Comment by rjeberhard** [comment]
@russgold, is there a test that captures how to handle existing services that don't have the serviceType label?  What will be the behavior?

**Comment by russgold** [comment]
That turns out to be a problem in the current system. What happened in the past is that a service with a cluster name was considered a cluster service, and one with a server name but no cluster name was be treated as a server service. The problem is, some server services _do_ have cluster names. That means that events regarding them have been putting them in the wrong place: in the clusters array rather than in the sko service field.

The new behavior reverses that: an old service with a server name is treated as a server service, while one with a cluster name but no server name is treated as a cluster service. I will add tests to validate that.

But since this version also includes validation by recipe checksum, all services in an existing install will be replaced, and the new ones will have the service type annotation.

File: operator/src/main/java/oracle/kubernetes/operator/helpers/DomainPresenceInfo.java [hunk 1]
```diff
@@ -71,28 +71,95 @@ private ServerKubernetesObjects getSko(String serverName) {
...
+        clusters,
+        clusterName,
+        s -> !KubernetesUtils.isFirstNewer(getMetadata(s), getMetadata(event)));
+  }
+
+  private static <K, V> boolean removeIfPresentAnd(
+      ConcurrentMap<K, V> map, K key, Predicate<? super V> predicateFunction) {
+    Objects.requireNonNull(predicateFunction);
+    for (V oldValue; (oldValue = map.get(key)) != null; ) {
+      if (!predicateFunction.test(oldValue)) return false;
```
Comment (markxnelson) [inline operator/src/main/java/oracle/kubernetes/operator/helpers/DomainPresenceInfo.java]: brackets

File: operator/src/main/java/oracle/kubernetes/operator/helpers/DomainPresenceInfo.java [hunk 2]
```diff
@@ -71,28 +71,95 @@ private ServerKubernetesObjects getSko(String serverName) {
...
-    getSko(serverName).setExternalService(service);
+    getSko(serverName).getExternalService().set(service);
+  }
+
+  void setExternalServiceFromEvent(String serverName, V1Service event) {
+    getSko(serverName).getExternalService().accumulateAndGet(event, this::getNewerService);
+  }
+
+  boolean deleteExternalServiceFromEvent(String serverName, V1Service event) {
+    if (serverName == null) return false;
```
Comment (markxnelson) [inline operator/src/main/java/oracle/kubernetes/operator/helpers/DomainPresenceInfo.java]: {}

File: operator/src/main/java/oracle/kubernetes/operator/helpers/KubernetesUtils.java [hunk 3]
```diff
@@ -141,4 +142,26 @@ static V1ObjectMeta getResourceMetadata(Object resource) {
...
+
+  /**
+   * Returns true if the first metadata indicates a newer resource than does the second.
+   *
+   * @param first the first item to compare
+   * @param second the second item to compare
+   * @return true if the first object is newer than the second object
+   */
+  public static boolean isFirstNewer(V1ObjectMeta first, V1ObjectMeta second) {
+    if (second == null) return true;
```
Comment (markxnelson) [inline operator/src/main/java/oracle/kubernetes/operator/helpers/KubernetesUtils.java]: {}

File: operator/src/main/java/oracle/kubernetes/operator/helpers/KubernetesUtils.java [hunk 4]
```diff
@@ -141,4 +142,26 @@ static V1ObjectMeta getResourceMetadata(Object resource) {
...
+    if (first == null) return false;
+
+    DateTime time1 = first.getCreationTimestamp();
+    DateTime time2 = second.getCreationTimestamp();
+
+    if (time2.isAfter(time1)) {
+      return false;
+    }
+    return Integer.parseInt(second.getResourceVersion())
```
Comment (markxnelson) [inline operator/src/main/java/oracle/kubernetes/operator/helpers/KubernetesUtils.java]: i would prefer this to be written so the intent is clear and i dont have to rtfm to work out what it does

File: operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java [hunk 5]
```diff
@@ -220,192 +216,26 @@ public void dispatchPodWatch(Watch.Response<V1Pod> item) {
...
-    for (V oldValue; (oldValue = map.get(key)) != null; ) {
-      if (predicateFunction.test(oldValue)) {
-        if (map.remove(key, oldValue)) {
-          return true;
-        }
-      } else {
-        return false;
-      }
+    switch (item.type) {
+      case "ADDED":
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/DomainProcessorImpl.java]: Huh. I find this to be a really common case, and clear when there is nothing between the cases.

File: operator/src/main/java/oracle/kubernetes/operator/helpers/KubernetesUtils.java (hunk 4 above)
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/KubernetesUtils.java]: Agreed, and will rewrite.

File: operator/src/main/java/oracle/kubernetes/operator/helpers/KubernetesUtils.java [hunk 6]
```diff
@@ -141,4 +102,33 @@ static V1ObjectMeta getResourceMetadata(Object resource) {
...
+    DateTime time1 = first.getCreationTimestamp();
+    DateTime time2 = second.getCreationTimestamp();
+
+    if (time1.isAfter(time2)) {
+      return true;
+    } else if (time2.isAfter(time1)) {
+      return false;
+    } else {
+      return getResourceVersion(first) > getResourceVersion(second);
+    }
```
Comment (tbarnes-us) [inline operator/src/main/java/oracle/kubernetes/operator/helpers/KubernetesUtils.java]: The original version you had for isNewer compares version numbers even if the timestamp of time2 'isAfter' time1 or time2 equals time1, and returns true if if the second version number is higher.   The new code only compares version numbers if time1 equals time2.   Is this intentional?

In other words, here's the logic of the new code (simplified a little vs. what you have above):

if (time1.isEquals(time2)) {
  return getResourceVersion(first) > getResourceVersion(second);
} else {
  return time1.isAfter(time2);
}

And here's the logic of the old code you had before:

if (time2.isAfter(time1)) {
  return false;
} else if (getResourceVersion(second) < getResourceVersion(first)) {
  return true;
} else {
  return time2.isBefore(time1);
}

Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/KubernetesUtils.java]: Yes, thank you. That is an improvement on what I had, and does exactly the same thing. Resource versions are just a tie-breaker for when the times are equal.

File: operator/src/test/java/oracle/kubernetes/operator/helpers/KubernetesUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
+// Copyright 2019, Oracle Corporation and/or its affiliates.  All rights reserved.
+// Licensed under the Universal Permissive License v 1.0 as shown at
+// http://oss.oracle.com/licenses/upl.
+
+package oracle.kubernetes.operator.helpers;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.*;
```
Comment (markxnelson) [inline operator/src/test/java/oracle/kubernetes/operator/helpers/KubernetesUtilsTest.java]: generally we do not allow wildcard imports, though this particular one is probably ok

**Pull request merged**